Control access to Assembly::JxWNeighbor

lindsayad opened the pull request:
Use the PassKey pattern to protect access of JxWNeighbor. This API
was developed for NodeFaceConstraints but users may be tempted to
incorrectly use this for things like DGKernels or InterfaceKernrels.
So using PassKey we can limit this to be accessible to
NodeFaceConstraint.

Closes #14870

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/481513/) on 41148f1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14890/site)

This comment will be updated on new commits.

lindsayad committed:
PassKey pattern to protect Assembly::JxWNeighbor

Use the PassKey pattern to protect access of JxWNeighbor. This API
was developed for NodeFaceConstraints but users may be tempted to
incorrectly use this for things like DGKernels or InterfaceKernrels.
So using PassKey we can limit this to be accessible to
NodeFaceConstraint.

Closes #14890

lindsayad commented:
I expect this to break SAM, which appears to be our only tested application which uses `JxWNeighbor`. It appears that it was accessing the API in a derived `NodeFaceConstraint` which is the appropriate conceptual use. They  probably had to do this because we didn't actually use `JxWNeighbor` in `NodeFaceConstraint` prior to this PR. This will require a patch to SAM since friendship is not inherited (and for a good reason; we should not trust users in general).

lindsayad commented:
I have the patch ready for SAM
